vsintegration/src/FSharp.Editor/CodeFix/AddMissingFunKeyword.fs

```diff
@@ -0,0 +1,67 @@
+﻿// Copyright (c) Microsoft Corporation.  All Rights Reserved.  See License.txt in the project root for license information.
+
+namespace Microsoft.VisualStudio.FSharp.Editor
+
+open System
+open System.Composition
+
+open Microsoft.CodeAnalysis.Text
+open Microsoft.CodeAnalysis.CodeFixes
+
+open FSharp.Compiler.SourceCodeServices
+
+[<ExportCodeFixProvider(FSharpConstants.FSharpLanguageName, Name = "AddMissingFunKeyword"); Shared>]
+type internal FSharpAddMissingFunKeywordCodeFixProvider
+    [<ImportingConstructor>]
+    (
+        projectInfoManager: FSharpProjectOptionsManager
+    ) =
+    inherit CodeFixProvider()
+
+    static let userOpName = "AddMissingFunKeyword"
+
+    let fixableDiagnosticIds = set ["FS0010"]
+
+    override _.FixableDiagnosticIds = Seq.toImmutableArray fixableDiagnosticIds
+
+    override _.RegisterCodeFixesAsync context =
+        asyncMaybe {
+            let document = context.Document
+            let! sourceText = context.Document.GetTextAsync(context.CancellationToken)
+            let textOfError = sourceText.GetSubText(context.Span).ToString()
+
+            // Only trigger when failing to parse `->`, which arises when `fun` is missing
+            do! Option.guard (textOfError = "->")
+
+            let! parsingOptions, _ = projectInfoManager.TryGetOptionsForEditingDocumentOrProject(document, context.CancellationToken, userOpName)
+            let defines = CompilerEnvironment.GetCompilationDefinesForEditing parsingOptions
+
+            let adjustedPosition =
+                let rec loop str pos =
+                    if not (String.IsNullOrWhiteSpace(str)) then
+                        pos
+                    else
+                        loop (sourceText.GetSubText(TextSpan(pos - 1, 1)).ToString()) (pos - 1)
+
+                loop (sourceText.GetSubText(TextSpan(context.Span.Start - 1, 1)).ToString()) context.Span.Start
```

Comment:
Is it possible that the construction of TextSpan would receive any invalid input? Such as, `context.Span.Start - 1` , would ever be `-1`?

Reply:
Not at this point, no. You can of course write literally `->` as the first set of characters in a script file, but it won't trigger this error when that happens. The compiler thinks it's in the middle of a `yield` expression to be precise, which is weird, but unrelated to this or any other code fixer.

Reply:
Sorry - it _does_ trigger FS0010, but the span isn't actually `->`, so it never reaches here in the first place